### PR TITLE
add missing potential_query_instability for keys and values in hashmap

### PR DESCRIPTION
**Pull request by chenyukang**
From https://github.com/rust-lang/rust/pull/120435#discussion_r1468883787,

These API are also returning iterator, so we need add `potential_query_instability` for them?

### CODE REVIEW

**Comment by chenyukang** [comment]
link #84447 #89558

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs [hunk 1]
```diff
@@ -403,6 +403,7 @@ fn codegenned_and_inlined_items(tcx: TyCtxt<'_>) -> DefIdSet {
     let mut result = items.clone();
 
     for cgu in cgus {
+        #[allow(rustc::potential_query_instability)]
```
Comment (michaelwoerister) [inline compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:406]: Can you try turning the `items` field in `CodegenUnit` into an `FxIndexMap`. That should make iteration stable (if the collection is built in a deterministic order). An `UnordMap` would be even better, but I suspect that that would be more work.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 2]
```diff
@@ -682,6 +682,7 @@ fn link_dwarf_object<'a>(
         }
 
         // Input rlibs contain .o/.dwo files from dependencies.
+        #[allow(rustc::potential_query_instability)]
         let input_rlibs = cg_results
             .crate_info
             .used_crate_source
```
Comment (michaelwoerister) [inline compiler/rustc_codegen_ssa/src/back/link.rs:688]: Can you turn the `used_crate_source` field into an `UnordMap`? Then you should be able to do this:
```rust
// Input rlibs contain .o/.dwo files from dependencies.
let input_rlibs = cg_results
    .crate_info
    .used_crate_source
    .items()
    .filter_map(|(_, csource)| csource.rlib.as_ref())
    .map(|(path, _)| path)
    .into_sorted_stable_ord();
```

File: compiler/rustc_hir_typeck/src/expr.rs [hunk 3]
```diff
@@ -1950,6 +1950,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     ) {
         let len = remaining_fields.len();
 
+        #[allow(rustc::potential_query_instability)]
```
Comment (michaelwoerister) [inline compiler/rustc_hir_typeck/src/expr.rs:1953]: It should be straightforward to make `remaining_fields` and `UnordMap`.

**Review by michaelwoerister — COMMENTED** [review]
Thanks for the PR, @chenyukang!

That's a great catch about the missing query instability attributes on the HashMap methods. That was definitely a bug. 

I've left a few comments below where it should not be too hard to fix the problem in a cleaner way.

**Comment by chenyukang** [comment]
> Thanks for the PR, @chenyukang!
> 
> That's a great catch about the missing query instability attributes on the HashMap methods. That was definitely a bug.
> 
> I've left a few comments below where it should not be too hard to fix the problem in a cleaner way.

There is a issue to track these kind of cleanups:
https://github.com/rust-lang/rust/issues/84447

I may create a separate PR to clean up those replacements?

**Comment by michaelwoerister** [comment]
This is also closely related: https://github.com/rust-lang/compiler-team/issues/533

It is where the [unord collections](https://doc.rust-lang.org/stable/nightly-rustc/rustc_data_structures/unord/index.html) were proposed. They've been implemented meanwhile and are quite mature (although it seems that `UnordMap` misses `fn keys(&self) -> UnordItems<&K, impl Iterator<Item = &K>>` for example). 

> I may create a separate PR to clean up those replacements?

Sure, you can r? @michaelwoerister on that. Let me know if you plan to do any further changes to this PR, otherwise I'm happy to approve it.

File: compiler/rustc_lint/src/context.rs [hunk 4]
```diff
@@ -326,11 +326,9 @@ impl LintStore {
...
     /// True if this symbol represents a lint group name.
     pub fn is_lint_group(&self, lint_name: Symbol) -> bool {
-        debug!(
-            "is_lint_group(lint_name={:?}, lint_groups={:?})",
-            lint_name,
-            self.lint_groups.keys().collect::<Vec<_>>()
-        );
+        #[allow(rustc::potential_query_instability)]
+        let lint_groups = self.lint_groups.keys().collect::<Vec<_>>();
+        debug!("is_lint_group(lint_name={:?}, lint_groups={:?})", lint_name, lint_groups);
```
Comment (klensy) [inline compiler/rustc_lint/src/context.rs:331]: Will collect things on release now, instead of debug only, next one too.

Comment (chenyukang) [reply compiler/rustc_lint/src/context.rs:331]: you are right, I will clean this up when I replace HashMap.

**Pull request merged**